### DISCUSSION TITLE
markdown code blocks add shell commands

### DISCUSSION THREAD

**Discussion by marek-step** [jgm/pandoc#10184] [Q&A] [answered]
Is it possible to add  commands to the pandoc list of shell-commands for code blocks? 

I am writing markdown texts with inserts in `{.tcsh}` code blocks. I compile them with pandoc over XeLaTeX to pdf. Pandoc translates the commands with `\KeywordTok{rm}` for example in the intermediate LaTeX translation process. These commands are rendered in the pdf in green colour. But some commands are rendered in black for example with `\NormalTok{systemextensionsctl}`. I tried to edit directly the code block in the original markdown file to no avail. 

For that reason my question: is it possible, to edit Pandoc somewhere, to add some shell commands? Like `pkgutil` `softwareupdate` or `grep`?

Thx

Edit: 

Here a screenshot for the missing shell commands. I insert the code blocks with:

```
<3backtics>{.zsh}
code coming here
<3backtics>
```

Right? 

<img width="637" alt="Screenshot 2024-09-15 at 11 53 04" src="https://github.com/user-attachments/assets/c1ac70e6-c129-4bf9-a347-9d515f30910c">

New Edit:

The in-between step from Markdown to Pdf is in Pandoc-LaTeX-Dialect as follows. Here you may see, that the commands are not coloured as they should. And editing this code block in Pandoc-LaTeX-Dialect as a `<3backtics>{=latex}` code block is not working (?)

```
\begin{Shaded}
  \begin{Highlighting}[]
  \ExtensionTok{systemextensionsctl}\NormalTok{ list}
    ...snip

  \FunctionTok{sudo}\NormalTok{ rm }\AttributeTok{{-}r} \StringTok{"/Library/Application Support/Cisco"}
  \FunctionTok{sudo}\NormalTok{ rm }\AttributeTok{{-}r}\NormalTok{ /opt/cisco}
  \FunctionTok{sudo}\NormalTok{ rm }\AttributeTok{{-}r}\NormalTok{ /Applications/Cisco}
  \FunctionTok{sudo}\NormalTok{ rm }\AttributeTok{{-}r}\NormalTok{ \textasciitilde{}/.cisco}
  \FunctionTok{sudo}\NormalTok{ rm }\AttributeTok{{-}r}\NormalTok{ \textasciitilde{}/.vpn}
  \ExtensionTok{pkgutil} \AttributeTok{{-}{-}pkgs} \KeywordTok{|} \FunctionTok{grep} \AttributeTok{{-}i}\NormalTok{ cisco}
  \FunctionTok{sudo}\NormalTok{ pkgutil }\AttributeTok{{-}{-}forget}\NormalTok{ com.cisco.pkg.anyconnect.vpn}
  \end{Highlighting}
  \end{Shaded}
```

**Comment by jgm** [comment]
See the manual for `--syntax-definition`.  You'd have to customize the syntax definition and pass it in with this option.

**Comment by bpj** [comment, marked as answer]
Not all highlighting schemes (or hoghlight-styles as Pandoc calls them) color all token types. You may want to use another highlighting style. Read the [manual] first.

  [manual]: https://pandoc.org/MANUAL.html#syntax-highlighting

If none of the builtin highlight styles is to your liking your best choice is to define youur own highlighting style. Print out the builtin one you want to modify with

``` shell
pandoc --print-highlight-style=STYLE > mystyle.json 
```

where STYLE is “pygments” or whatever you choose from the builtin ones.

When you open the JSON file you will see something like this

``` json
{
    "text-color": null,
    "background-color": null,
    "line-number-color": "#aaaaaa",
    "line-number-background-color": null,
    "text-styles": {
        "Alert": {
            "text-color": "#ff0000",
            "background-color": null,
            "bold": true,
            "italic": false,
            "underline": false
        },
```

where “Alert” and several others following it are token types. It is self-explanatory: you can change the the text color, background color and font styles of tokens. If you are like me you will need some actual colors to look at rather than experimenting with changing hex digita; the [X11 colors page on Wikipedia] is a good place to start.

  [X11 colors page on Wikipedia]: https://en.m.wikipedia.org/wiki/X11_color_names

I don’t know if it is possible but I assume that if some token type in the marked-up code is missing in the style definition file I assume you can just add it.

To use your custom style just use pandoc’s `--highlight-style=FILENAME` option (and include the path if it isn’t in the current directory.)

**Reply by marek-step** [reply]
thx to jgm and bpj for your answers. 

This is opening a new aspect of pandoc, which I am using over 5 years now. Looking for one style `pygments` with `pandoc --print-highlight-style=pygments > mystyle.json` I don't see a definition for the shell-commands, only a general for `Keywords`. I was hoping for a style file, where I could add the missing shell commands, like `rm` `pkgutil` etc 
I have to dig further into it with the helpful link to the pandoc-manual. If it is possible, I will add a screenshot into my original posting, to illustrate my question better.

**Reply by tarleb** [reply]
See [here](https://kate-editor.org/syntax/) for syntax definition files, including those for tcsh and zsh. You can download and edit the respective `.xml` files and pass them to pandoc via `--syntax-definition`.

**Reply by marek-step** [reply]
Finally I realised, that my question was wrong. I was mislead by the green colour of `sudo`, thinking all shell commands should be green too. But `sudo` is not a command - isn't it? - and the shell commands `rm` `pkgutil` etc are already "coloured"; in black! In any case your answers opened for me a new facet to the powerful Pandoc. Thanks again

**Reply by tarleb** [reply]
No, you were right, known unix commands are highlighted in colors other than black.

I forgot to mention in my earlier message that `tcsh.xml` depends on the definitions in `bash.xml`, so you might have to download and pass both those files.

**Reply by marek-step** [reply]
thx tarleb, you are right: the commands are not coloured. I tried to download the definitions from the link you suggested, but the file formats are in `.html`. I don't  found a solution, to pass these files to `pandoc` with `--syntax-definition`.
I will edit my question again, to show the pandoc translation of the shell-code block to LaTeX before translating it to pdf. Here you see clearly, that the shell-commands are not discovered.

**Reply by bpj** [reply]
Go to their [repository](https://github.com/KDE/syntax-highlighting/tree/master/data/syntax) and choose the Download item in the menu to the upper right of the file content display for each file.

**Reply by bpj** [reply]
Also see the [handbook](https://docs.kde.org/stable5/en/kate/katepart/highlight.html) and links from there. (Scroll down a bit for the part about syntax definition files)

Disclaimer: I found creating or even modifying these files hard, but then I don't like XML as a data format...

**Reply by marek-step** [reply]
Finally I made a big progress: I could insert the bash.xml into my Makefile `--syntax-definition    ./bash.xml  \`. I could add some missing commands into the KDE File: `<item>systemextensionsctl</item>` or `<item>pkgutil</item>` ... And it is coloured as expected. 
One little deception: after `sudo` the defined commands are not coloured : `rm` or `pkgutil`
Little remark again: editing the Pandoc-LaTeX as shown in my question and passing it as a Code Block with `<3backtics>{=latex}` is possible. Probably this will be easier for people like me, which don't want to dig into KDE-XML
